python/ray/data/dataset.py

```diff
@@ -3469,6 +3478,76 @@ def write_fn_wrapper(blocks: Iterator[Block], ctx, fn) -> Iterator[Block]:
             datasource.on_write_failed([], e)
             raise
 
+    @ConsumptionAPI(pattern="Time complexity:")
+    def write_datasink(
+        self,
+        datasink: Datasink,
+        *,
+        ray_remote_args: Dict[str, Any] = None,
+    ) -> None:
+        """Writes the dataset to a custom :class:`~ray.data.Datasink`.
+
+        Time complexity: O(dataset size / parallelism)
+
+        Args:
+            datasink: The :class:`~ray.data.Datasink` to write to.
+            ray_remote_args: Kwargs passed to ``ray.remote`` in the write tasks.
+        """  # noqa: E501
+        if ray_remote_args is None:
+            ray_remote_args = {}
+
+        if not datasink.supports_distributed_writes:
+            if ray.util.client.ray.is_connected():
+                raise ValueError(
+                    "If you're using Ray Client, Ray Data won't schedule write tasks "
+                    "on the driver's node."
+                )
+            ray_remote_args["scheduling_strategy"] = NodeAffinitySchedulingStrategy(
+                ray.get_runtime_context().get_node_id(),
+                soft=False,
+            )
+
```

Comment:
Is this method the same as `Dataset.write_datasource()` except the `supports_distributed_writes`? Can we create a common function to avoid repeating same code?

Reply:
No. We pass in `**write_args` to `Write` in `write_datasource`, but we don't do so in `write_datasink`.

Given that we're hard deprecating `write_datasource` in a few weeks, I think it's okay if we need leave these decoupled.